### PR TITLE
core: Add lexer for more complex pass pipeline specs

### PR DESCRIPTION
**Pull request by AntonLydike**
This PR starts to add pass arguments. We first provide a very simple tokenizer for the syntax of passes. We implement a subset of MLIRs pass syntax here.

This is needed to get passes such as domain slicing for MPI going.

### CODE REVIEW

**Review by webmiche — APPROVED** [review]
LGTM.

File: xdsl/utils/parse_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+def tokenize_pass(input_str: str) -> Iterator[Token]:
+    """
+    This tokenizes a pass declaration string. Pass syntax is a subset
+    of MLIRs pass pipeline syntax:
+
+    pipeline          ::= pipeline-element (`,` pipeline-element)*
+    pipeline-element  ::= pass-name options?
+    options           ::= `{` options-element ( ` ` options-element)* `}`
+    options-element   ::= key `=` value
```
Comment (PapyChacal) [inline xdsl/utils/parse_pipeline.py:50]: Could we have
```
options-element ::= key ( `=` value )
```
here? It is used in MLIR and useful at times!

Comment (AntonLydike) [reply xdsl/utils/parse_pipeline.py:50]: I was considering that, but I am unsure what the value should be in that case. Boolean?

I would do that in a later PR, as I fear this to be a bit of a discussion point...

**Comment by AntonLydike** [comment]
Closed in favor of the existing lexer!

**Comment by AntonLydike** [comment]
Aaaand here we are again! Re-opened as we can't really use the existing lexer.

File: xdsl/utils/parse_pipeline.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    """
+    This tokenizes a pass declaration string. Pass syntax is a subset
+    of MLIRs pass pipeline syntax:
+    pipeline          ::= pipeline-element (`,` pipeline-element)*
+    pipeline-element  ::= pass-name options?
+    options           ::= `{` options-element ( ` ` options-element)* `}`
+    options-element   ::= key (`=` value (`,` value)* )?
+    key       ::= IDENT
+    pass-name ::= IDENT
+    value     :== NUMBER / BOOL / IDENT / STRING_LITERAL
```
Comment (math-fehr) [inline xdsl/utils/parse_pipeline.py:52]:
```suggestion
    value     ::= NUMBER | BOOL | IDENT | STRING_LITERAL
```

File: xdsl/utils/parse_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        self._peeked = None
+        return token
+
+    def peek(self) -> Token:
+        if self._peeked is None:
+            self._peeked = next(self._stream)
+        return self._peeked
+
+
+class PassPipelineParseError(BaseException):
```
Comment (math-fehr) [inline xdsl/utils/parse_pipeline.py:95]: Should we move that to our `exception` file?

File: xdsl/utils/parse_pipeline.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        end = len(input_str)
+
+        while True:
+            token: Token | None = None
+            for pattern, kind in _lexer_rules:
+                if (match := pattern.match(input_str, pos)) is not None:
+                    token = Token(Span(match.start(), match.end(), input), kind)
+                    pos = match.end()
+                    break
+            if token is None:
```
Comment (math-fehr) [inline xdsl/utils/parse_pipeline.py:75]: This could be a use for the `for/else` statement ;)

File: xdsl/utils/parse_pipeline.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
+import re
+from dataclasses import dataclass
+from typing import Iterator
+from xdsl.utils.lexer import Input, Span
+from enum import Enum, auto
+
+
+class Kind(Enum):
+    EOF = auto()
```
Comment (math-fehr) [inline xdsl/utils/parse_pipeline.py:9]: nit: If you want to be sure no one is using the values you get from `auto`, you can use `object()` instead.

File: xdsl/utils/parse_pipeline.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
+import re
+from dataclasses import dataclass
+from typing import Iterator
+from xdsl.utils.lexer import Input, Span
+from enum import Enum, auto
+
+
+class Kind(Enum):
```
Comment (math-fehr) [inline xdsl/utils/parse_pipeline.py:8]: I would rename that `TokenKind`? Or put it inside the `Token`.

**Review by math-fehr — APPROVED** [review]
:+1:

File: xdsl/utils/parse_pipeline.py (hunk 4 above)
Comment (AntonLydike) [reply xdsl/utils/parse_pipeline.py:75]: I thought about it, but I think that `for/else` is not a feature I want to *add* to a codebase^^

File: xdsl/utils/parse_pipeline.py (hunk 1 above)
Comment (PapyChacal) [reply xdsl/utils/parse_pipeline.py]: Oh okay, I don't expect it but better keep this first one as clear as possible sure

File: xdsl/utils/parse_pipeline.py (hunk 4 above)
Comment (math-fehr) [reply xdsl/utils/parse_pipeline.py:75]: That's perfectly understandable ;)

File: xdsl/utils/parse_pipeline.py (hunk 1 above)
Comment (AntonLydike) [reply xdsl/utils/parse_pipeline.py]: We do have this now :D

**Comment by AntonLydike** [comment]
@math-fehr I think I addressed all your points now!

**Pull request merged**